File: backend/internal/domain/facility/field_device.go
package facility

import (
	"github.com/besart951/go_infra_link/backend/internal/domain"
	"github.com/google/uuid"
)

type FieldDevice struct {
	domain.Base
	BMK                       *string
	Description               *string
	ApparatNr                 int
	TextIndividuell           *string
	SPSControllerSystemTypeID uuid.UUID               `gorm:"type:uuid;not null;index"`
	SPSControllerSystemType   SPSControllerSystemType `gorm:"foreignKey:SPSControllerSystemTypeID;constraint:OnUpdate:CASCADE,OnDelete:CASCADE;"`
	SystemPartID              uuid.UUID               `gorm:"type:uuid;not null;index"`
	SystemPart                SystemPart              `gorm:"foreignKey:SystemPartID"`
	SpecificationID           *uuid.UUID              `gorm:"type:uuid;index"`
	Specification             *Specification          `gorm:"foreignKey:SpecificationID;constraint:OnUpdate:CASCADE,OnDelete:SET NULL;"`
	ApparatID                 uuid.UUID               `gorm:"type:uuid;not null;index"`
	Apparat                   Apparat                 `gorm:"foreignKey:ApparatID"`

	BacnetObjects []BacnetObject `gorm:"foreignKey:FieldDeviceID"`
}

// FieldDeviceOptions contains all metadata needed for creating/editing field devices
type FieldDeviceOptions struct {
	Apparats            []Apparat
	SystemParts         []SystemPart
	ObjectDatas         []ObjectData
	ApparatToSystemPart map[uuid.UUID][]uuid.UUID // apparat_id -> [system_part_ids]
	ObjectDataToApparat map[uuid.UUID][]uuid.UUID // object_data_id -> [apparat_ids]
}

// FieldDeviceFilterParams contains optional filter parameters for listing field devices
type FieldDeviceFilterParams struct {
	BuildingID                *uuid.UUID
	ControlCabinetID          *uuid.UUID
	SPSControllerID           *uuid.UUID
	SPSControllerSystemTypeID *uuid.UUID
	ProjectID                 *uuid.UUID
	ProjectIDs                []uuid.UUID
}

// FieldDeviceCreateItem represents a single field device to create in a multi-create operation
type FieldDeviceCreateItem struct {
	FieldDevice   *FieldDevice
	ObjectDataID  *uuid.UUID
	BacnetObjects []BacnetObject
}

// FieldDeviceCreateResult represents the result of creating a single field device
type FieldDeviceCreateResult struct {
	Index       int          // Index in the original request array
	Success     bool         // Whether the creation succeeded
	FieldDevice *FieldDevice // The created field device (nil if failed)
	Error       string       // Error message if failed (empty if succeeded)
	ErrorField  string       // Specific field that caused the error (if applicable)
}

// FieldDeviceMultiCreateResult represents the result of a multi-create operation
type FieldDeviceMultiCreateResult struct {
	Results       []FieldDeviceCreateResult
	TotalRequests int
	SuccessCount  int
	FailureCount  int
}

// BulkFieldDeviceUpdate represents a single field device update in a bulk operation
type BulkFieldDeviceUpdate struct {
	ID                 uuid.UUID
	BMK                *string
	HasBMK             bool
	Description        *string
	HasDescription     bool
	TextIndividuell    *string
	HasTextIndividuell bool
	ApparatNr          *int
	ApparatID          *uuid.UUID
	SystemPartID       *uuid.UUID
	Specification      *SpecificationPatch
	BacnetObjects      *[]BacnetObjectPatch
}

func (u BulkFieldDeviceUpdate) HasBMKUpdate() bool {
	return u.HasBMK || u.BMK != nil
}

func (u BulkFieldDeviceUpdate) HasDescriptionUpdate() bool {
	return u.HasDescription || u.Description != nil
}

func (u BulkFieldDeviceUpdate) HasTextIndividuellUpdate() bool {
	return u.HasTextIndividuell || u.TextIndividuell != nil
}

type SpecificationPatch struct {
	SpecificationSupplier                        *string
	HasSpecificationSupplier                     bool
	SpecificationBrand                           *string
	HasSpecificationBrand                        bool
	SpecificationType                            *string
	HasSpecificationType                         bool
	AdditionalInfoMotorValve                     *string
	HasAdditionalInfoMotorValve                  bool
	AdditionalInfoSize                           *int
	HasAdditionalInfoSize                        bool
	AdditionalInformationInstallationLocation    *string
	HasAdditionalInformationInstallationLocation bool
	ElectricalConnectionPH                       *int
	HasElectricalConnectionPH                    bool
	ElectricalConnectionACDC                     *string
	HasElectricalConnectionACDC                  bool
	ElectricalConnectionAmperage                 *float64
	HasElectricalConnectionAmperage              bool
	ElectricalConnectionPower                    *float64
	HasElectricalConnectionPower                 bool
	ElectricalConnectionRotation                 *int
	HasElectricalConnectionRotation              bool
}

func (p *SpecificationPatch) HasChanges() bool {
	if p == nil {
		return false
	}
	return p.HasSpecificationSupplier ||
		p.HasSpecificationBrand ||
		p.HasSpecificationType ||
		p.HasAdditionalInfoMotorValve ||
		p.HasAdditionalInfoSize ||
		p.HasAdditionalInformationInstallationLocation ||
		p.HasElectricalConnectionPH ||
		p.HasElectricalConnectionACDC ||
		p.HasElectricalConnectionAmperage ||
		p.HasElectricalConnectionPower ||
		p.HasElectricalConnectionRotation
}

func (p *SpecificationPatch) HasNonNilValues() bool {
	if p == nil {
		return false
	}
	return p.SpecificationSupplier != nil ||
		p.SpecificationBrand != nil ||
		p.SpecificationType != nil ||
		p.AdditionalInfoMotorValve != nil ||
		p.AdditionalInfoSize != nil ||
		p.AdditionalInformationInstallationLocation != nil ||
		p.ElectricalConnectionPH != nil ||
		p.ElectricalConnectionACDC != nil ||
		p.ElectricalConnectionAmperage != nil ||
		p.ElectricalConnectionPower != nil ||
		p.ElectricalConnectionRotation != nil
}

// BulkOperationResultItem represents the result of a single item in a bulk operation
type BulkOperationResultItem struct {
	ID      uuid.UUID
	Success bool
	Error   string
	Fields  map[string]string // Per-field validation errors (e.g., "bacnet_objects.0.text_fix" -> "message")
}

// BulkOperationResult represents the result of a bulk operation
type BulkOperationResult struct {
	Results      []BulkOperationResultItem
	TotalCount   int
	SuccessCount int
	FailureCount int
}

File: frontend/src/lib/domain/facility/field-device.ts
/**
 * Field Device domain types
 * Mirrors backend: internal/domain/facility/field_device.go
 */

import type { Pagination } from '../utils/index.ts';
import type { Apparat } from './apparat.js';
import type { BacnetObject } from './bacnet-object.js';
import type { Specification } from './specification.js';
import type { SPSControllerSystemType } from './sps-sys-type.js';
import type { SystemPart } from './system.js';

export interface FieldDevice {
	id: string;
	bmk?: string;
	description?: string;
	apparat_nr: string;
	sps_controller_system_type_id: string;
	system_part_id?: string;
	specification_id?: string;
	apparat_id: string;
	created_at: string;
	updated_at: string;

	// Embedded related entities for display
	sps_controller_system_type?: SPSControllerSystemType;
	apparat?: Apparat;
	system_part?: SystemPart;
	specification?: Specification;
	bacnet_objects?: BacnetObject[];
}

export interface BacnetObjectInput {
	text_fix: string;
	description?: string;
	gms_visible: boolean;
	optional: boolean;
	text_individual?: string;
	software_type: string;
	software_number: number;
	hardware_type: string;
	hardware_quantity: number;
	software_reference_id?: string;
	state_text_id?: string;
	notification_class_id?: string;
	alarm_definition_id?: string;
}

export interface CreateFieldDeviceRequest {
	bmk?: string;
	description?: string;
	apparat_nr: number;
	sps_controller_system_type_id: string;
	system_part_id: string;
	apparat_id: string;
	object_data_id?: string;
	bacnet_objects?: BacnetObjectInput[];
}

export interface UpdateFieldDeviceRequest {
	bmk?: string;
	description?: string;
	apparat_nr?: number;
	sps_controller_system_type_id?: string;
	system_part_id?: string;
	apparat_id?: string;
	object_data_id?: string;
	bacnet_objects?: BacnetObjectInput[];
}

export interface FieldDeviceListParams {
	page?: number;
	limit?: number;
	search?: string;
	sps_controller_system_type_id?: string;
}

export interface FieldDeviceListResponse extends Pagination<FieldDevice> {}

/**
 * FieldDeviceOptions - Response from /api/v1/facility/field-devices/options
 * Contains all metadata needed for creating/editing field devices with relationships
 */
export interface FieldDeviceOptions {
	apparats: import('./apparat.js').Apparat[];
	system_parts: import('./system.js').SystemPart[];
	object_datas: import('./object-data.js').ObjectData[];
	apparat_to_system_part: Record<string, string[]>; // apparat_id -> [system_part_ids]
	object_data_to_apparat: Record<string, string[]>; // object_data_id -> [apparat_ids]
}

/**
 * Available apparat numbers response
 */
export interface AvailableApparatNumbersResponse {
	available: number[];
}

/**
 * Multi-create field device request
 */
export interface MultiCreateFieldDeviceRequest {
	field_devices: CreateFieldDeviceRequest[];
}

/**
 * Result for a single field device creation in multi-create
 */
export interface FieldDeviceCreateResult {
	index: number; // Index in the original request array
	success: boolean; // Whether the creation succeeded
	field_device?: FieldDevice; // The created field device (null if failed)
	error: string; // Error message if failed (empty if succeeded)
	error_field: string; // Specific field that caused the error (if applicable)
}

/**
 * Multi-create field device response
 */
export interface MultiCreateFieldDeviceResponse {
	results: FieldDeviceCreateResult[];
	total_requests: number;
	success_count: number;
	failure_count: number;
}

/**
 * Bulk update field device item
 */
export interface BulkUpdateFieldDeviceItem {
	id: string;
	bmk?: string;
	description?: string;
	apparat_nr?: number;
	apparat_id?: string;
	system_part_id?: string;
}

/**
 * Bulk update field device request
 */
export interface BulkUpdateFieldDeviceRequest {
	updates: BulkUpdateFieldDeviceItem[];
}

/**
 * Bulk operation result item
 */
export interface BulkOperationResultItem {
	id: string;
	success: boolean;
	error?: string;
}

/**
 * Bulk update field device response
 */
export interface BulkUpdateFieldDeviceResponse {
	results: BulkOperationResultItem[];
	total_count: number;
	success_count: number;
	failure_count: number;
}

/**
 * Bulk delete field device request
 */
export interface BulkDeleteFieldDeviceRequest {
	ids: string[];
}

/**
 * Bulk delete field device response
 */
export interface BulkDeleteFieldDeviceResponse {
	results: BulkOperationResultItem[];
	total_count: number;
	success_count: number;
	failure_count: number;
}
